### PR TITLE
docs: add op counter bilingual tutorial

### PR DESCRIPTION
**Pull request by AllenYolk**
## Summary
- add bilingual CN/EN tutorials for `activation_based.op_counter`
- cover `DispatchCounterMode`, core counters, roofline workflow, and four high-level energy estimators
- refine onboarding examples from a new-user perspective while keeping the tutorial in SNN context

## Included
- overview of `op_counter` and counter workflow
- roofline example using `FlopCounter` and `MemoryAccessCounter`
- boundaries for `compute_energy`, `Lemaire`, `NeuroMC`, and `SpikeSim`
- inference energy examples for `compute-only` and `Lemaire`

## Validation
- `cd docs && make html` succeeded in the worktree
- the repo still has pre-existing import/build warnings and an existing `IndentationError` in `spikingjelly/activation_based/cuda_kernel/neuron_kernel/integrate_and_fire.py`, which blocks direct example execution through normal imports

## Summary by CodeRabbit

* **Documentation**
  * Added new English and Chinese "Advanced" tutorials on runtime operator profiling and energy estimation, now listed in each TOC.
  * Tutorials cover runtime operator counting (FLOPs, memory accesses, SynOps, MACs, ACs), example workflows including roofline-style analysis, and practical guidance on choosing/reporting estimators.
  * Minor spacing adjustment in legacy tutorials section.

[Review Change Stack](https://app.coderabbit.ai/change-stack/fangwei123456/spikingjelly/pull/683?utm_source=github_walkthrough&utm_medium=github&utm_campaign=change_stack)

### CODE REVIEW

**Review by AllenYolk — COMMENTED** [review]
## Code Review Summary

**Verdict: Comment** — Well-structured documentation, a few minor observations.

### Suggestions

- **EN op_counter.rst:12-13** — List items start with lowercase (`count`, `build`). Consider capitalizing for consistency with standard English list formatting (`Count`, `Build`).

### Looks Good

- Clear, well-organized bilingual tutorial covering the full op_counter workflow from basic counters to high-level energy estimators.
- Consistent structure between CN and EN versions.
- Practical guidance on choosing the right estimator for the task is helpful for new users.
- Cross-references (`class`, `doc`) are correctly formatted.

### Verified

- CN and EN tutorials are structurally symmetric.
- Section heading hierarchy is consistent (document =, section +, subsection -) in both language versions.
- All code examples appear syntactically valid.

---
*Reviewed by Hermes Agent*

**Comment by AllenYolk** [comment]
## Code Review Summary — PR #683

**Verdict: COMMENT** (good work — no blocking issues, 2 minor observations)

### ✅ Verified Correct

**API accuracy** — All code examples cross-checked against the actual module source on `master`:

- `FlopCounter`, `MemoryAccessCounter`, `SynOpCounter`, `MACCounter`, `ACCounter` — correct class names and method signatures (`get_total()`, `get_counts()["Global"]`)
- `DispatchCounterMode([...], verbose, strict)` — parameters match `base.py`
- `estimate_compute_energy(model, x, config=cfg)` — keyword-only `config` parameter ✓
- `ComputeEnergyConfig(cost_config=ComputeEnergyCostConfig.fp16())` — class chain correct
- `report.energy_total_pj`, `.energy_mac_pj`, `.energy_ac_pj`, `.counts` — match `ComputeEnergyReport`
- `estimate_lemaire_energy` → `total_pj`, `breakdown_pj` — match `LemaireEnergyReport`
- `estimate_neuromc_runtime_energy` / `estimate_spikesim_event_energy` — exist in `__init__.py` exports

**Index files** — Both `cn/index.rst` (line 34) and `en/index.rst` (line 34) correctly add `op_counter` under the Advanced section, in alphabetical position.

**Roofline example clarity** — Line 135 explicitly says *"If you only care about inference roofline, remove the backward() call"* — good documentation practice.

**RST formatting** — Heading underlines all match their title lengths. Toctree entries follow project conventions.

### Code Quality

- Clean, well-structured bilingual content with parallel section ordering
- Clear separation between basic counters and high-level energy estimators
- The estimator comparison table is well-organized and correctly states each estimator's boundaries
- The **"Do not compare absolute numbers across different estimators"** warning is appropriately emphasized

### Minor Observations

1. **`model.eval()` is only mentioned in the energy estimator section** (line 311) but not in the basic counting workflow. For the simple models used in the basic examples (Linear + IFNode) this is harmless, but consider adding a brief note in the workflow section for users who copy-paste with stateful layers.

2. **The SynOp section** could briefly note that if a model uses ReLU between conv layers (as the roofline example does), SynOps will be zero for those layers since ReLU produces dense float outputs. Not a bug, but helpful context for new users.

### ✅ Looks Good

- No secrets, debug artifacts, or tool state files in the diff
- Both CN and EN versions are structurally consistent
- Code blocks are properly formatted and executable as-is
- Representative input shapes and sparsity caveats are well-stated

---

*Reviewed by Hermes Agent*

**Comment by AllenYolk** [comment]
@coderabbitai review\n@gemini-code-assist review

**Comment by AllenYolk** [comment]
@coderabbitai review\n@gemini-code-assist review

**Comment by AllenYolk** [comment]
@coderabbitai review\n@gemini-code-assist review

**Comment by AllenYolk** [comment]
## Code Review Summary

**Verdict: COMMENT** (self-review — author is AllenYolk)

---

### ✅ Looks Good

**Thorough API verification.** All code examples correctly reference real API surface:
- `FlopCounter`, `MemoryAccessCounter`, `SynOpCounter`, `MACCounter`, `ACCounter` — correct class names and method signatures (`get_total()`, `get_counts()["Global"]`)
- `DispatchCounterMode([...], verbose, strict)` — parameters match `base.py`
- `estimate_compute_energy(model, x, config=cfg)` — keyword-only `config` correctly used
- `ComputeEnergyConfig(cost_config=ComputeEnergyCostConfig.fp16())` — class chain correct
- `report.energy_total_pj`, `.energy_mac_pj`, `.energy_ac_pj`, `.counts` — match `ComputeEnergyReport`
- `estimate_lemaire_energy` → `total_pj`, `breakdown_pj` — match `LemaireEnergyReport`

**RST formatting is valid.** `make html` succeeds.

**`nn.ReLU()` correctly removed** from the roofline example (commit `7f97fa0` ✅).

**Clear estimator boundaries.** The "do not compare across estimators" warning is prominent.

**Good pedagogical flow.** Overview → basic counting → roofline → energy estimators → practical examples → guidance.

**CN version extras.** The `graph LR` diagram and tool-selection guide are nice additions.

---

### ⚠️ Warnings

**1. `strict=True` with backward pass in roofline example** `(Correctness)`

Both EN and CN roofline examples use `strict=True` with `y.sum().backward()`:
```python
with op_counter.DispatchCounterMode([flop_counter, mem_counter], strict=True):
    y = model(x)
    y.sum().backward()
```

The backward pass dispatches backward-specific ops (`aten.convolution_backward`). If `FlopCounter`/`MemoryAccessCounter` don't register handlers for these, `strict=True` will raise `NotImplementedError`. The PR description says this was fixed by changing to `strict=False`, but the current file at `4ed0a2b` still has `strict=True`.

**Action:** Run this exact example to verify it doesn't crash. If it does, either change to `strict=False` or remove the backward pass from this example.

**2. Missing `model.eval()` in basic counting examples** `(Robustness)`

The basic counting workflow example doesn't use `.eval()`. While harmless for the specific model (`Linear` + `IFNode` + `Linear`), users copying this pattern for models with stateful layers (BatchNorm, dropout) may get incorrect counts. The compute-energy example correctly uses `.eval()` — consider being consistent.

---

### 💡 Suggestions

**1. Redundant `requires_grad_(True)`** `(Code Quality)`

In the roofline example:
```python
for p in model.parameters():
    p.requires_grad_(True)
```
`nn.Conv2d` parameters already have `requires_grad=True` by default. This is a no-op. Consider removing or adding a comment explaining why it's included.

**2. Bilingual content sync** `(Documentation)`

The CN version includes a `graph LR` architecture diagram, tool-selection guide, reporting standards, and recommended workflow — ensure the EN version has parallel sections.

**3. Consider adding `.eval()` to basic workflow** `(Documentation)`

Adding `.eval()` to the basic counting example would make the tutorial more robust for copy-paste users.

---

### 🔍 Verified Correct

| Concern | Status |
|---------|--------|
| Basic counting example with `strict=True`, forward only | ✅ All dispatched ops should be handled |
| SynOpCounter with binary spike input | ✅ Handles `aten.addmm` on binary input |
| Roofline (no ReLU, forward only) | ✅ Conv2d ops handled |
| `ComputeEnergyConfig.fp16()` class method | ✅ Exists in module |
| Lemaire report fields `total_pj`, `breakdown_pj` | ✅ Match `LemaireEnergyReport` |

---
*Reviewed by Hermes Agent*

**Comment by AllenYolk** [comment]
@coderabbitai review\n@gemini-code-assist review
